### PR TITLE
Some Windows fixes

### LINKED ISSUE #36: Builds failing on windows
Currently builds are failing on windows: https://github.com/pestphp/pest/runs/704328906?check_suite_focus=true.

Screenshot 2020-05-28 at 01 37 22

Some help would be appreciated here.

### PR DESCRIPTION
**Pull request by Pierstoval**
Binary was using realpath, and targets were sanitized without taking the full path's drive name in account, so this was fixed.

I intend to wait for CI to make sure the test suite still works as expected.

I hope it'll fix #36 🤔

### CODE REVIEW

**Comment by nunomaduro** [comment]
Screenshot 2020-05-29 at 20 11 56

**Comment by nunomaduro** [comment]
@Pierstoval Can you rebase buddy? The fix of the build is now on the master.

**Comment by nunomaduro** [comment]
Yup, couple of tests on windows are failing. This is good sign.

**Comment by Pierstoval** [comment]
Locally, Pest tests work, but i have some that fail:

```
   WARN  EdevwwwpesttestsFeaturesSkip
  ✓ it do not skips
  s it skips with truthy
  s it skips with truthy condition by default
  s it skips with message → skipped because bar
  s it skips with truthy closure condition
  ✓ it do not skips with falsy closure condition
  s it skips with condition and message → skipped because foo

   FAIL  EdevwwwpesttestsPHPUnitCustomTestCaseInSubFoldersSubFolderSubFolderUsesPerSubDirectory
  • closure was bound to  custom test case

   FAIL  EdevwwwpesttestsPHPUnitCustomTestCaseUsesPerDirectory
  • closure was bound to  custom test case

   FAIL  EdevwwwpesttestsPluginsTraits
  • it allows global uses

   FAIL  EdevwwwpesttestsVisualSingleTestOrDirectory
  • allows to run a single test
  • allows to run a directory
```

(sorry, output is not really easy to copy to github 😅 )

**Comment by owenvoke** [comment]
Yeah, I was trying to work those out earlier. 🤔 (https://github.com/pestphp/pest/pull/41#pullrequestreview-420547668)

**Comment by Pierstoval** [comment]
Some tests are related to the fact that `php` isn't called, and `bin/pest`, I'm investigating :)

**Comment by Pierstoval** [comment]
Another issue is how test names are displayed in the output, I've got this issue after a fix:

```
 $ php bin/pest tests\Visual\SingleTestOrDirectory.php

   FAIL  EdevwwwpesttestsVisualSingleTestOrDirectory
  • allows to run a single test
  • allows to run a directory

  ---

  • EdevwwwpesttestsVisualSingleTestOrDirectory > allows to run a single test
  Failed asserting that '\n
     PASS  TestsFixturesDirectoryWithTestsExampleTest\r\n
    ✓ it example\r\n
  \n
    Tests:  1 passed\r\n
    Time:   0.04s\r\n
  ' contains "   PASS  Tests\Fixtures\DirectoryWithTests\ExampleTest
    ✓ it example

    Tests:  1 passed".
```

As you can see, the test name should be `Tests\Fixtures\DirectoryWithTests\ExampleTest`, but it's displayed as `TestsFixturesDirectoryWithTestsExampleTest`.

I'm trying to search where this could be done... @nunomaduro, any idea?

**Comment by nunomaduro** [comment]
@Pierstoval Here: https://github.com/nunomaduro/collision/blob/a4ef05049dbefcacef5081ce34943dc4ef8106fa/src/Adapters/Phpunit/TestResult.php#L113.

**Comment by Pierstoval** [comment]
I made a few changes, it's still investigation more than bug fixes since I'm only testing this on Windows.

I'm totally unsure of 6fb3c79 since I don't know what's going to be the impact on Linux 🤔 but I was unsure about using `basename()` since it didn't display the full test name, instead was just displaying the class name and not its FQN in the test output.

By default, with `basename()`, I get this:

```
$ php bin/pest tests\Visual\SingleTestOrDirectory.php

   FAIL  SingleTestOrDirectory
  • allows to run a single test
  • allows to run a directory

  ---

  • SingleTestOrDirectory > allows to run a single test
  Failed asserting that '\n
     PASS  ExampleTest\r\n
    ✓ it example\r\n
  \n
    Tests:  1 passed\r\n
    Time:   0.04s\r\n
  ' contains "   PASS  Tests\Fixtures\DirectoryWithTests\ExampleTest
    ✓ it example

    Tests:  1 passed".
```


If I replace `basename()` with the `preg_replace()` in 6fb3c79 , I get this:

```
$ php bin/pest tests\Visual\SingleTestOrDirectory.php

   FAIL  Tests\Visual\SingleTestOrDirectory
  • allows to run a single test
  • allows to run a directory

  ---

  • Tests\Visual\SingleTestOrDirectory > allows to run a single test
  Failed asserting that '\n
     PASS  E\dev\www\pest\tests\Fixtures\DirectoryWithTests\ExampleTest\r\n
    ✓ it example\r\n
  \n
    Tests:  1 passed\r\n
    Time:   0.04s\r\n
  ' contains "   PASS  Tests\Fixtures\DirectoryWithTests\ExampleTest
    ✓ it example

    Tests:  1 passed".
```

There seem to be tons of things related to paths and root dir, I'm not sure where to search next🤔

**Comment by Pierstoval** [comment]
(marked it as draft because it seems it's definitely **NOT** ready to be merged)

File: src/Factories/TestCaseFactory.php [hunk 1]
```diff
@@ -152,15 +152,23 @@ public function build(TestSuite $testSuite): array
      */
     public function makeClassFromFilename(string $filename): string
     {
+        if ('\\' === DIRECTORY_SEPARATOR) {
```
Comment (nunomaduro) [inline src/Factories/TestCaseFactory.php:155]: This will only happen on windows?

Comment (owenvoke) [reply src/Factories/TestCaseFactory.php:155]: Might be better to just use `PHP_OS_FAMILY === 'Windows'`?

Comment (Pierstoval) [reply src/Factories/TestCaseFactory.php:155]: Yep this can happen only on Windows, that's a fairly common case 🙂

**Comment by nunomaduro** [comment]
@Pierstoval Hey buddy, you are still working on this correct? Its fine if takes weeks, I just wanna make sure that there is no action on my side to be done.

**Comment by Pierstoval** [comment]
For now, I need to check why tests are failing, but I'm a bit busy these days :)

**Comment by nunomaduro** [comment]
No problem. Going to mark this issue as `needs help`. Maybe someone can take over.

**Comment by dimitrioskarvounaris** [comment]
So, what exactly is here more required? Can you elaborate a little more?

**Comment by nunomaduro** [comment]
Sure, normally if you are on windows you can reproduce the problem running: `composer test:unit`.

**Comment by Pierstoval** [comment]
@dimitrioskarvounaris  The tests must be fully working on Windows, which is not the case now

**Comment by dimitrioskarvounaris** [comment]
Continuing work under https://github.com/dimitrioskarvounaris/pest/tree/windows-gitbash
I got the pull request with the github-cli and then tried to push it to my repo to not create separate branches and maybe add it on top of this one, but it will keep asking permissions to Pierstoval's repo to commit on the same branch, even though remote url points to my fork. If you know a workaround to this, let me know.

**Comment by dimitrioskarvounaris** [comment]
Felt it was too much trouble, to have one more fork (if that would even work at all, to make a fork from a fork, while having another fork of the original repo), and github wouldn't let me push to the same branch even on my fork, as I used the github-cli to clone the pull request... sorry... so I just made a new PR here: https://github.com/pestphp/pest/pull/61 - it contains though all previous commits too

**Comment by nunomaduro** [comment]
This is the branch that contains the windows work? If yes, can you close the other one @dimitrioskarvounaris?

**Comment by nunomaduro** [comment]
I've removed windows for now: https://github.com/pestphp/pest/commit/e8d426f57406ce0ad0638f88557aa4aaa3b00bd8.

Fell free to put it back.

**Comment by dimitrioskarvounaris** [comment]
@nunomaduro Branch "windows-gitbash" contains Pierstoval's as also my fixes, not "windows". https://github.com/pestphp/pest/pull/61fixes everything and the tests pass (I run them on windows as also linux).

**Comment by nunomaduro** [comment]
Closing in favour of https://github.com/pestphp/pest/pull/61.

**Comment by sgzapata** [comment]
Dear colleagues, I am working on software engineering research and request your cooperation in answering this survey: https://forms.gle/h2D6cfjJM6rtSV456

**It will only take you 10 minutes.**

Excuse me, **thank you very much for your help**.

**Comment by dkarvounaris** [comment]
> 
> 
> Dear colleagues, I am working on software engineering research and request your cooperation in answering this survey: https://forms.gle/h2D6cfjJM6rtSV456
> 
> **It will only take you 10 minutes.**
> 
> Excuse me, **thank you very much for your help**.

Is this some spam?

**Comment by sgzapata** [comment]
No Dimitrios, it is not spam.
I am a PhD candidate researching software engineering on Github software teams.
Excuse me, **I hope you can cooperate with me. Thanks.**

**Pull request closed without merging**